common/http_test.go

```diff
@@ -0,0 +1,171 @@
+package common_test
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"reflect"
+	"strings"
+	"sync"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/tendermint/tmlibs/common"
+)
+
+func TestWriteSuccess(t *testing.T) {
+	w := httptest.NewRecorder()
+	common.WriteSuccess(w, "foo")
+	assert.Equal(t, w.Code, 200, "should get a 200")
+}
+
+var blankErrResponse = new(common.ErrorResponse)
+
+func TestWriteError(t *testing.T) {
+	w := httptest.NewRecorder()
+	msg := "this is an error"
+	common.WriteError(w, errors.New(msg))
+	assert.Equal(t, w.Code, http.StatusBadRequest, "should get a 400")
+	blob, err := ioutil.ReadAll(w.Body)
+	require.Equal(t, err, nil, "expecting a successful ioutil.ReadAll")
+
+	recv := new(common.ErrorResponse)
+	err = json.Unmarshal(blob, recv)
+	require.Equal(t, err, nil, "expecting a successful json.Unmarshal")
+
+	require.Equal(t, reflect.DeepEqual(recv, blankErrResponse), false, "expecting a non-blank error response")
+}
+
+func TestWriteCode(t *testing.T) {
+	codes := []int{
+		http.StatusOK,
+		http.StatusBadRequest,
+		http.StatusUnauthorized,
+		http.StatusInternalServerError,
+	}
+
+	for _, code := range codes {
+		w := httptest.NewRecorder()
+		common.WriteCode(w, "foo", code)
+		assert.Equal(t, w.Code, code, "expecting the same code recorded")
+	}
+}
+
+type saver struct {
+	Foo int    `json:"foo" validate:"min=10"`
+	Bar string `json:"bar"`
+}
+
+type rcloser struct {
+	closeOnce sync.Once
+	body      *bytes.Buffer
+	closeChan chan bool
+}
+
+var errAlreadyClosed = errors.New("already closed")
+
+func (rc *rcloser) Close() error {
+	var err = errAlreadyClosed
+	rc.closeOnce.Do(func() {
+		err = nil
+		rc.closeChan <- true
+		close(rc.closeChan)
+	})
+	return err
+}
+
+func (rc *rcloser) Read(b []byte) (int, error) {
+	return rc.body.Read(b)
+}
+
+var _ io.ReadCloser = (*rcloser)(nil)
+
+func makeReq(strBody string) (*http.Request, <-chan bool) {
+	closeChan := make(chan bool, 1)
+	buf := new(bytes.Buffer)
+	buf.Write([]byte(strBody))
+	req := &http.Request{
+		Header: make(http.Header),
+		Body:   &rcloser{body: buf, closeChan: closeChan},
+	}
+	return req, closeChan
+}
+
+func TestParseRequestJSON(t *testing.T) {
+	tests := [...]struct {
+		body    string
+		want    interface{}
+		wantErr bool
+	}{
```

Comment:
want is always nil? shouldn't you provide a `new(InputStruct)` or something like that.  Otherwise, of course they all error. It would be good to see some positive results as well.